Feature/calling server v3 events

yyw-msft opened the pull request:
- events added with serialization methods
- abstract participants related events into internal events and public facing events
- added EventParser class to help parse calling server call back events

yyw-msft committed:
events added, serialization missing

yyw-msft committed:
adding events and serializations

yyw-msft committed:
updated participants related events

yyw-msft committed:
update transfer events

yyw-msft committed:
fix addParticipantsSucceededEvent

yyw-msft committed:
Merge branch 'main' of github.com:Azure/azure-sdk-for-net into feature/callingServerV3Events

yyw-msft committed:
adding api doc

minwoolee-msft commented on sdk/communication/Azure.Communication.CallingServer/src/Models/Events/AcsEventType.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace Azure.Communication.CallingServer
+{
+    public partial struct AcsEventType
```
This is overwriting generated AcsEventType right? it is difficult to tell which is overwriting, and which is standalone model. Can you add [CodeGenModel("??")] for all overwriting models?

yyw-msft replied:
added!

minwoolee-msft commented on sdk/communication/Azure.Communication.CallingServer/src/Models/Events/CallingServerEventBase.cs:
```diff
@@ -0,0 +1,12 @@
+    /// <summary>
+     /// The CallingServer Event Base.
+     /// </summary>
+    public partial class CallingServerEventBase
```
All Events seems to have 1 common property - AcsEventType - is it possible to include that in here in base class?
the benefit is that the SDK user can easily use base classe's EventType to figure out which event it parsed into

yyw-msft replied:
Added AcsEventType to the EventBase class. I had the field non-nullable instead of using `AcsEventType?` do u foresee any problem with that? My reasoning is that we should always construct an Event with an AcsEventType. Do you see this being too restrictive?

minwoolee-msft replied:
This is good - one of ACS review board mentioned, considering making this non-nullable. Thank you for pointing out.

jasonshave commented:
@yyw-msft could you update the description with a code sample of how a developer would use these enhancements? Also, in one of the branches for handling the publishing of events we have a struct called `AcsCallData` which wraps the three main properties we always give back. Here is an example:

```csharp
    public class CallConnectedEvent : AcsCallbackEvent
    {
        public CallConnectedEvent(AcsCallData acsCallData) 
            : base(acsCallData)
        {
        }
    }
```

The `AcsCallbackEvent` in this case consumes the immutable  `AcsCallData` and assigns the properties to the base class. Give me a ring if you want to discuss.

yyw-msft committed:
addressed PR comments

yyw-msft committed:
Merge branch 'main' of github.com:Azure/azure-sdk-for-net into feature/callingServerV3Events

yyw-msft committed:
updated api file

minwoolee-msft commented on sdk/communication/Azure.Communication.CallingServer/src/Models/Events/AddParticipantsFailedEvent.cs:
```diff
@@ -0,0 +1,64 @@
+        /// <summary> Participants failed to be added. </summary>
+        public IReadOnlyList<CommunicationIdentifier> Participants { get; }
+        /// <summary> Gets the type. </summary>
+        public AcsEventType? Type { get; }
```
Because now we have Base abstract class has EventType, this should go away. All derived events to remove Type.

yyw-msft replied:
updated all the derived event classes! thx for catching it.

yyw-msft committed:
removed type field from derived event classes, removed 'event' in event class names, removed hardcoded event type match

yyw-msft commented:
> @yyw-msft could you update the description with a code sample of how a developer would use these enhancements? Also, in one of the branches for handling the publishing of events we have a struct called `AcsCallData` which wraps the three main properties we always give back. Here is an example:
> 
> ```cs
>   public class CallConnectedEvent : AcsCallbackEvent
>   {
>       public CallConnectedEvent(AcsCallData acsCallData) 
>           : base(acsCallData)
>       {
>       }
>   }
> ```
> 
> The `AcsCallbackEvent` in this case consumes the immutable `AcsCallData` and assigns the properties to the base class. Give me a ring if you want to discuss.

@jasonshave Per our conversation offline, I have removed "event" from all the event classes name and added string manipulation on CloudEvent.Type.
